ga4gh/backends.py

```diff
@@ -368,99 +311,104 @@ def getMetadata(self):
-                response.nextPageToken = '{0}:'.format(variantSetIndex + 1)
-                # this token will give results from the beginning of the
-                # next variantSet
-        else:
-            response.nextPageToken = "{0}:{1}".format(variantSetIndex,
-                                                      response.nextPageToken)
-        return response
+    def runSearchRequest(
+            self, requestStr, requestClass, responseClass, pageListName,
+            objectGenerator):
+        """
+        Runs the specified request. The request is a string containing
+        a JSON representation of an instance of the specified requestClass
+        in which the page list variable has the specified pageListName.
+        We return a string representation of an instance of the specified
+        responseClass in JSON format. Objects are filled into the page list
+        using the specified object generator, which must return
+        (object, nextPageToken) pairs, and be able to resume iteration from
+        any point using the nextPageToken attribute of the request object.
+        """
+        request = requestClass.fromJSON(requestStr)
+        pageList = []
+        nextPageToken = None
+        for obj, nextPageToken in objectGenerator(request):
+            pageList.append(obj)
+            if len(pageList) >= request.pageSize:
+                break
+        response = responseClass()
+        response.nextPageToken = nextPageToken
+        setattr(response, pageListName, pageList)
+        return response.toJSON()
 
     def searchVariantSets(self, request):
         """
-        Returns a GASearchVariantSets response for the specified
+        Returns a GASearchVariantSetsResponse for the specified
         GASearchVariantSetsRequest object.
         """
-        # This is temporary until we do this properly based on the output
-        # buffer size
-        print(request)
-        if request.pageSize is None:
-            request.pageSize = 100
-        response = protocol.GASearchVariantSetsResponse()
+        return self.runSearchRequest(
+            request, protocol.GASearchVariantSetsRequest,
+            protocol.GASearchVariantSetsResponse, "variantSets",
+            self.variantSetsGenerator)
+
+    def searchVariants(self, request):
+        """
+        Returns a GASearchVariantsResponse for the specified
+        GASearchVariantsRequest  object.
+        """
+        return self.runSearchRequest(
+            request, protocol.GASearchVariantsRequest,
+            protocol.GASearchVariantsResponse, "variants",
+            self.variantsGenerator)
+
+    def variantSetsGenerator(self, request):
+        """
+        Returns a generator over the (variantSet, nextPageToken) pairs defined
+        by the speficied request.
+        """
         currentIndex = 0
         if request.pageToken is not None:
             currentIndex = int(request.pageToken)
```

Comment:
Not trying to rush anyone, but as with the previous round of cleanups, I'm
waiting for this refactoring to land before I implement the new htslib
VCF/BCF backend.

On Mon, Jan 12, 2015 at 7:32 AM, Jerome Kelleher <notifications@github.com>
wrote:

> In ga4gh/backends.py
> <https://github.com/ga4gh/server/pull/59#discussion-diff-22796944>:
>
> > +
> > +    def searchVariants(self, request):
> > +        """
> > +        Returns a GASearchVariantsResponse for the specified
> > +        GASearchVariantsRequest  object.
> > +        """
> > +        return self.runSearchRequest(
> > +            request, protocol.GASearchVariantsRequest,
> > +            protocol.GASearchVariantsResponse, "variants",
> > +            self.variantsGenerator)
> > +
> > +    def variantSetsGenerator(self, request):
> > +        """
> > +        Returns a generator over the (variantSet, nextPageToken) pairs defined
> > +        by the speficied request.
> > +        """
> >          currentIndex = 0
> >          if request.pageToken is not None:
> >              currentIndex = int(request.pageToken)
>
> It is in this case though. PageTokens are handled locally here, so that
> each generator is locally responsible for managing its own. So, for
> variantSets, the pageToken is always an int, and for variants, the
> pageToken is always a ":" seperated pair of ints. We'll have to do
> something a bit more complicated with these when we start supporting
> Datasets, but I think the extension should be easy enough.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ga4gh/server/pull/59/files#r22796944>.
>